### PR TITLE
extract detection of challenge/group tasks #8452

### LINKED ISSUE #8452: abstract the code that checks for a task belonging to a challenge or a group plan
As described in [this comment from paglias](https://github.com/HabitRPG/habitica/pull/8427#issuecomment-274513364) the challenge and group plans code created in https://github.com/HabitRPG/habitica/pull/8427 should be abstracted to a static method inside the Task model, so that duplication of the code can be removed.

@lynxlynxlynx has offered to work on this (thank you!).

Once it's finished, work can proceed on https://github.com/HabitRPG/habitica/pull/8213, which also needs to check whether tasks belong to a challenge or group plan.

### PR DESCRIPTION
**Pull request by lynxlynxlynx**
just a late refactor ...

Fixes #8452

### CODE REVIEW

File: website/server/models/task.js [hunk 1]
```diff
@@ -174,6 +174,21 @@ TaskSchema.methods.scoreChallengeTask = async function scoreChallengeTask (delta
   await chalTask.save();
 };
 
+// skips challenge and group tasks
+export let skipChallengeTasks = [
```
Comment (TheHollidayInn) [inline website/server/models/task.js]: Can you add the word qurey params to this? Maybe `skipChallengeTasksQuery`? And, I think this is technically a constant.

**Comment by TheHollidayInn** [comment]
I think the correct solution for this would be to abstract this function to a library or to the model:
```
let tasks = await Tasks.Task.find({
        userId: user._id,		        userId: user._id,
        type: {$in: ['daily', 'habit', 'todo']},		        type: {$in: ['daily', 'habit', 'todo']},
 -      $and: [ // exclude challenge and group tasks		 +      $and: Tasks.skipChallengeTasks,
 -        {		
 -          $or: [		
 -            {'challenge.id': {$exists: false}},		
 -            {'challenge.broken': {$exists: true}},		
 -          ],		
 -        },		
 -        {		
 -          $or: [		
 -            {'group.id': {$exists: false}},		
 -            {'group.broken': {$exists: true}},		
 -          ],		
 -        },		
 -      ],		
      }).exec();		      }).exec();
```

The function should be called something like `getNonGroupNonChallengeTasks`. This would correctly abstract the underlying database logic, and provide clarity.

Our next step would to create a custom Task query builder: `taskrepository.getTasks.excludeGroup().excludeChallenge.query()`

**Review by TheHollidayInn — COMMENTED** [review]
I think the correct solution for this would be to abstract this function to a library or to the model:
```
let tasks = await Tasks.Task.find({
        userId: user._id,		        userId: user._id,
        type: {$in: ['daily', 'habit', 'todo']},		        type: {$in: ['daily', 'habit', 'todo']},
 -      $and: [ // exclude challenge and group tasks		 +      $and: Tasks.skipChallengeTasks,
 -        {		
 -          $or: [		
 -            {'challenge.id': {$exists: false}},		
 -            {'challenge.broken': {$exists: true}},		
 -          ],		
 -        },		
 -        {		
 -          $or: [		
 -            {'group.id': {$exists: false}},		
 -            {'group.broken': {$exists: true}},		
 -          ],		
 -        },		
 -      ],		
      }).exec();		      }).exec();
```

The function should be called something like `getNonGroupNonChallengeTasks`. This would correctly abstract the underlying database logic, and provide clarity.

Our next step would to create a custom Task query builder: `taskrepository.getTasks.excludeGroup().excludeChallenge.query()`

File: website/server/models/task.js (hunk 1 above)
Comment (paglias) [reply website/server/models/task.js]: 👍

**Comment by librarianmage** [comment]
Any status on this PR? I am waiting to finish #8213, which requires this.

**Comment by lynxlynxlynx** [comment]
Sorry, forgot about this. I'll go refactor it into the proposed getNonGroupNonChallengeTasks method.

**Comment by lynxlynxlynx** [comment]
I did the refactoring of the find/remove calls, but I can't make it actually work. Nothing I tried in the method itself did it. I think I have the find invocation wrong?

**Comment by librarianmage** [comment]
@lynxlynxlynx Were you able to make it work?

**Comment by lynxlynxlynx** [comment]
No, I'm missing something, as the errors are useless. I've been lazy and just using the buildbot for tests, but this requires a proper inspection, so I'll have to set up full dev env.

**Comment by Alys** [comment]
@lynxlynxlynx It's important to always use a dev env for all PRs. All affected features always need careful manual testing to ensure they still work correctly. You can't rely solely on the automated tests.

**Comment by SabreCat** [comment]
@lynxlynxlynx, this PR has been awaiting your updates for a while. Are you still working on it?

**Comment by lynxlynxlynx** [comment]
Not currently, no. A quickfix would be to dig the constification+rename from the reflog, but that's all I can promise in this busy month.

**Pull request closed without merging**